File: src/main/java/com/nugurang/service/TeamService.java
package com.nugurang.service;

import com.nugurang.constant.RoleName;
import com.nugurang.dao.RoleDao;
import com.nugurang.dao.TeamDao;
import com.nugurang.dao.XrefUserTeamDao;
import com.nugurang.dto.TeamInputDto;
import com.nugurang.entity.TeamEntity;
import com.nugurang.entity.XrefUserTeamEntity;
import java.util.List;
import java.util.Optional;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class TeamService {
    private final UserService userService;
    private final RoleDao roleDao;
    private final TeamDao teamDao;
    private final XrefUserTeamDao xrefUserTeamDao;

    @Transactional
    public TeamEntity createTeam(TeamInputDto teamInputDto) {
        TeamEntity teamEntity = teamDao.save(TeamEntity.builder().name(teamInputDto.getName()).build());
        xrefUserTeamDao.save(XrefUserTeamEntity.builder().user(userService.getCurrentUser().get()).team(teamEntity).role(roleDao.findByName(RoleName.OWNER.name())).build());
        return teamEntity;
    }

    public Optional<TeamEntity> getTeam(Long teamId) {
        return teamDao.findById(teamId);
    }

    public Optional<TeamEntity> getTeam(String teamName) {
        return Optional.ofNullable(teamDao.findByName(teamName));
    }

    public List<TeamEntity> getTeams(String teamName, Pageable pageable) {
        return teamDao.findAllByNameContainingIgnoreCase(teamName, pageable).getContent();
    }

    public TeamEntity updateTeam(TeamInputDto teamInputDto, Long id) {
        TeamEntity teamEntity = teamDao.findById(id).get();
        teamEntity.setName(teamInputDto.getName());
        return teamDao.save(teamEntity);
    }

    public void deleteTeam(Long teamId) {
        teamDao.deleteById(teamId);
    }

    //<editor-fold defaultstate="collapsed" desc="delombok">
    @SuppressWarnings("all")
    
    public TeamService(final UserService userService, final RoleDao roleDao, final TeamDao teamDao, final XrefUserTeamDao xrefUserTeamDao) {
        this.userService = userService;
        this.roleDao = roleDao;
        this.teamDao = teamDao;
        this.xrefUserTeamDao = xrefUserTeamDao;
    }
    //</editor-fold>
}

File: src/main/kotlin/com/nugurang/service/TeamService.kt
package com.nugurang.service

import com.nugurang.annotation.DaoOp
import com.nugurang.constant.RoleName
import com.nugurang.dao.RoleDao
import com.nugurang.dao.TeamDao
import com.nugurang.dao.XrefUserTeamDao
import com.nugurang.dto.TeamInputDto
import com.nugurang.entity.RoleEntity
import com.nugurang.entity.TeamEntity
import com.nugurang.entity.XrefUserTeamEntity
import com.nugurang.exception.NotFoundException
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service

@Service
class TeamService(
    private val userService: UserService,
    private val roleDao: RoleDao,
    private val teamDao: TeamDao,
    private val xrefUserTeamDao: XrefUserTeamDao
) {
    @DaoOp
    fun createTeam(teamInputDto: TeamInputDto): TeamEntity {
        val teamEntity = teamDao.save(TeamEntity(name = teamInputDto.name))
        xrefUserTeamDao.save(
            XrefUserTeamEntity(
                user = userService.getCurrentUser(),
                team = teamEntity,
                role = roleDao.findByName(RoleName.OWNER.name) ?: throw NotFoundException(RoleEntity::class.java),
            )
        )
        return teamEntity
    }

    fun getTeam(teamId: Long): TeamEntity {
        return teamDao.findByIdOrNull(teamId) ?: throw NotFoundException(TeamEntity::class.java)
    }

    fun getTeam(teamName: String): TeamEntity {
        return teamDao.findByName(teamName) ?: throw NotFoundException(TeamEntity::class.java)
    }

    fun getTeams(teamName: String, pageable: Pageable): List<TeamEntity> {
        return teamDao.findAllByNameContainingIgnoreCase(teamName, pageable).content
    }

    @DaoOp
    fun updateTeam(teamInputDto: TeamInputDto, id: Long): TeamEntity {
        val teamEntity = teamDao.findByIdOrNull(id) ?: throw NotFoundException(TeamEntity::class.java)
        teamEntity.name = teamInputDto.name
        return teamDao.save(teamEntity)
    }

    @DaoOp
    fun deleteTeam(teamId: Long) {
        teamDao.deleteById(teamId)
    }
}